### PR TITLE
[PVT#181020089] Implement better deployment monitoring and try out asset precompilation in GAH

### PR DESCRIPTION
**Pull request by andyzito**
[Pivotal Story](https://www.pivotaltracker.com/n/projects/1548451/stories/181020089)

# Logging

I replaced our manual log tailing with a call out to the awscliv2's log tailing, which mimics `tail -f` and is more stable than our homerolled version. This has significantly reduced the need for asking the user to confirm continuation of the deploy task if the logs can't be found/aren't up to date.

# Monitoring

I just added basic timing to a bunch of deploy steps. Here are some times from a representative deployment:

|              STEP              | SECONDS |
| ------------------------------ | ------- |
| download secrets (.env)        | 2       |
| database materialize           | 1       |
| sync_app_assets                | 2       |
| clobbering assets              | 63      |
| checksumming assets            | 1       |
| ------------------------------ | ------- |
| pulling compiled assets        | 4       |
| rake maintenance:create        | 32      |
| rake db:migrate:primary        | 36      |
| rake db:migrate:warehouse      | 77      |
| rake db:migrate:health         | 70      |
| rake db:migrate:reporting      | 32      |
| rake reports:seed              | 40      |
| rake db:seed                   | 54      |
| rake gettext:sync_to_po_and_db | 269     |
| ./bin/cron_installer.rb        | 14      |

# Asset Precompilation in GitHub Actions

Here's the flow as it exists in this code:

1. Code is pushed
2. GitHub Action "asset_compilation" starts
  a. Pull in secrets file from Secrets Manager (parameter store char limit too low)
  b. Start bin/compile_assets.rb, which iterates through each install and...
    1. Pulls down the theme files from `s3://openpath-ecs-assets/#{target_group_name}`
    2. Pulls down the client .env secrets into .env.secrets
    3. Generates a checksum based on `app/assets` (including the downloaded theme files) and .env.secrets
    4. Checks if we have `s3://openpath-ecs-assets/#{target_group_name}/#{checksum}`
    5. If we don't, assets have changed. Run `assets:precompile` and push the results (`public/assets`) to `s3://openpath-ecs-assets/#{target_group_name}/#{checksum}`.
3. Deployment is triggered.
  a. Checksum is generated again, and if matching compiled assets are not available, the deploy will warn you and wait for 60 seconds to try again
5. Eventually, after the deploy proceeds, an application container is spun up. In the entrypoint...
  a. Pull down theme files and generate checksum again (this doesn't seem to be very computationally expensive)
  b. Pull down compiled assets the same way we do now, except with the checksum instead of the githash

The idea is that we are directly mapping the contents of the uncompiled assets with the compiled assets in S3, instead of mapping via the githash. And so it will always be easy to check if we have a valid compiled version for the current asset code.

### CODE REVIEW

File: .github/workflows/asset_compilation.yml [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+        aws-region: us-east-1
+
+    - name: Set up dependencies
+      env:
+        DATABASE_APP_DB_TEST: boston_hmis_test
+        WAREHOUSE_DATABASE_DB_TEST: warehouse_test
+        HEALTH_DATABASE_DB_TEST: health_test
+        DATABASE_DB_TEST: boston_cas_test
+        REPORTING_DATABASE_DB_TEST: test_boston_reporting
```
Comment (eanders) [inline .github/workflows/asset_compilation.yml]: These can probably just be:
```suggestion
        DATABASE_DB_TEST: cas_test
        REPORTING_DATABASE_DB_TEST: reporting_test
```

File: .github/workflows/asset_compilation.yml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        HEALTH_FROM: greenriver.testing@mailinator.com
+        # HOSTNAME: openpath.host
+        # PORT: 80
+        REPORTING_DATABASE_ADAPTER: postgresql
+        # REPORTING_DATABASE_DB_TEST: test_boston_reporting
+        REPORTING_DATABASE_HOST: postgres
+        REPORTING_DATABASE_PASS: postgres
+        REPORTING_DATABASE_USER: postgres
+        WAREHOUSE_DATABASE_ADAPTER: postgis
+        # WAREHOUSE_DATABASE_DB_TEST: warehouse_test
```
Comment (eanders) [inline .github/workflows/asset_compilation.yml]:
```suggestion
        DATABASE_ADAPTER: postgresql
        DATABASE_HOST: postgres
        DATABASE_PASS: postgres
        DATABASE_USER: postgres
        DEFAULT_FROM: greenriver.testing@mailinator.com
        ENCRYPTION_KEY: strongEncryptionstrongEncryptionstrongEncryption
        HEALTH_DATABASE_ADAPTER: postgresql
        HEALTH_DATABASE_HOST: postgres
        HEALTH_DATABASE_PASS: postgres
        HEALTH_DATABASE_USER: postgres
        HEALTH_FROM: greenriver.testing@mailinator.com
        REPORTING_DATABASE_ADAPTER: postgresql
        REPORTING_DATABASE_HOST: postgres
        REPORTING_DATABASE_PASS: postgres
        REPORTING_DATABASE_USER: postgres
        WAREHOUSE_DATABASE_ADAPTER: postgis
```

File: .github/workflows/audit.yml [hunk 3]
```diff
@@ -1,52 +0,0 @@
-name: Bundle Audit and Brakeman
```
Comment (eanders) [inline .github/workflows/audit.yml]: This will need to be restored, right?

File: .github/workflows/rails_tests.yml [hunk 4]
```diff
@@ -1,199 +0,0 @@
-name: Rails Tests
```
Comment (eanders) [inline .github/workflows/rails_tests.yml]: This will need to be restored, right?

Comment (andyzito) [reply .github/workflows/rails_tests.yml]: Yes.

File: .github/workflows/audit.yml (hunk 3 above)
Comment (andyzito) [reply .github/workflows/audit.yml]: Yes.

**Comment by andyzito** [comment]
@eanders Per my most recent commit message, `File.sanitize` isn't builtin and I think w/[the other sanitization stuff I put in](https://github.com/greenriver/hmis-warehouse/pull/2157/commits/358961b3c1962c3338ed4592ac5a77b996fce04b) it's probably fine? But lmk if you're still not feeling good about it, I assume we can add `File.sanitize` without too much trouble.

**Comment by eanders** [comment]
> @eanders Per my most recent commit message, `File.sanitize` isn't builtin and I think w/[the other sanitization stuff I put in](https://github.com/greenriver/hmis-warehouse/pull/2157/commits/358961b3c1962c3338ed4592ac5a77b996fce04b) it's probably fine? But lmk if you're still not feeling good about it, I assume we can add `File.sanitize` without too much trouble.

I think what you have is fine.  Thanks!

**Pull request merged**